modules/javafx.graphics/src/main/java/com/sun/javafx/css/media/MediaQuerySerializer.java

```diff
@@ -0,0 +1,117 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.sun.javafx.css.media;
+
+import com.sun.javafx.css.media.expression.ConjunctionExpression;
+import com.sun.javafx.css.media.expression.ConstantExpression;
+import com.sun.javafx.css.media.expression.FunctionExpression;
+import com.sun.javafx.css.media.expression.NegationExpression;
+import com.sun.javafx.css.media.expression.DisjunctionExpression;
+import javafx.css.StyleConverter;
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+/**
+ * Serializes and deserializes a {@link MediaQuery} expression into and from a binary representation.
+ */
+public final class MediaQuerySerializer {
+
+    private MediaQuerySerializer() {}
+
+    /**
+     * The order of {@code QueryType} enum values is important!
+     * It is part of the binary serialization protocol for media query expressions.
+     */
+    private enum QueryType {
+        CONSTANT,
+        FUNCTION,
+        CONJUNCTION,
+        DISJUNCTION,
+        NEGATION;
+
+        static QueryType of(MediaQuery expression) {
+            return switch (expression) {
+                case ConstantExpression _ -> CONSTANT;
+                case FunctionExpression<?> _ -> FUNCTION;
+                case ConjunctionExpression _ -> CONJUNCTION;
+                case DisjunctionExpression _ -> DISJUNCTION;
+                case NegationExpression _ -> NEGATION;
+            };
+        }
+
+        static final QueryType[] VALUES = values();
+    }
+
+    public static void writeBinary(MediaQuery mediaQuery,
+                                   DataOutputStream os,
+                                   StyleConverter.StringStore stringStore) throws IOException {
+        os.writeByte(QueryType.of(mediaQuery).ordinal());
+
+        switch (mediaQuery) {
+            case ConstantExpression expr ->
+                os.writeBoolean(expr.value());
+
+            case FunctionExpression<?> expr -> {
+                os.writeInt(stringStore.addString(expr.featureName()));
+
+                if (expr.featureValue() != null) {
+                    os.writeInt(stringStore.addString(expr.featureValue()));
+                } else {
+                    os.writeInt(-1);
+                }
+            }
+
+            case NegationExpression expr ->
+                writeBinary(expr.expression(), os, stringStore);
+
+            case ConjunctionExpression expr -> {
+                writeBinary(expr.left(), os, stringStore);
+                writeBinary(expr.right(), os, stringStore);
+            }
+
+            case DisjunctionExpression expr -> {
+                writeBinary(expr.left(), os, stringStore);
+                writeBinary(expr.right(), os, stringStore);
+            }
+        }
+    }
+
+    public static MediaQuery readBinary(DataInputStream is, String[] strings) throws IOException {
+        return switch (QueryType.VALUES[is.readByte()]) {
```

Comment:
same issue with backward compatibility - what if an old core tries to read the newer format and AIOOBE is thrown instead of IOException?  Should we validate first and throw IOE instead?

Reply:
That would be forward compatibility, which is not supported. JavaFX will not even try to load a binary file with a higher version than it can understand, and error out immediately (see Stylesheet.java:L317).